### PR TITLE
[AMDGPU][GFX12] Default component broadcast store

### PR DESCRIPTION
**Pull request by mariusz-sikora-at-amd**
For image and buffer stores the default behaviour on GFX12 is to set all unset components to the value of the first component. So if we pass only X component, it will be the same as XXXX, or XY same as XYXX.

This patch simplifies the passed vector of components in InstCombine by removing components from the end that are equal to the first component.

For image stores it also trims DMask if necessary.

### CODE REVIEW

File: llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-simplify-image-buffer-stores.ll [hunk 1]
```diff
@@ -23,7 +23,8 @@ define amdgpu_ps void @image_store_1d_store_insert_zeros_at_end(<8 x i32> inreg
 ; GCN-NEXT:    ret void
 ;
 ; GFX12-LABEL: @image_store_1d_store_insert_zeros_at_end(
-; GFX12-NEXT:    call void @llvm.amdgcn.image.store.1d.f32.i32(float [[VDATA1:%.*]], i32 1, i32 [[S:%.*]], <8 x i32> [[RSRC:%.*]], i32 0, i32 0)
+; GFX12-NEXT:    [[NEWVDATA4:%.*]] = insertelement <4 x float> <float poison, float 0.000000e+00, float 0.000000e+00, float 0.000000e+00>, float [[VDATA1:%.*]], i64 0
+; GFX12-NEXT:    call void @llvm.amdgcn.image.store.1d.v4f32.i32(<4 x float> [[NEWVDATA4]], i32 15, i32 [[S:%.*]], <8 x i32> [[RSRC:%.*]], i32 0, i32 0)
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-simplify-image-buffer-stores.ll:27]: This resulted in using a larger register type, which is worse?

Comment (jayfoad) [reply llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-simplify-image-buffer-stores.ll:27]: Right. The behaviour changed in GFX12, so the old trailing zeros optimization no longer applies (which is what this diff shows) but the new broadcast optimization applies instead.

Comment (arsenm) [reply llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-simplify-image-buffer-stores.ll:27]: Seems like an unfortunate change. Is the old optimization *wrong*? I have a lot more questions and concerns if so. We can't let optimizations run on IR without a set target-cpu break a program that ultimately executes on gfx12. Do we need an optimization that unfolds the scalar form to the vector with broadcast?

Comment (jayfoad) [reply llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-simplify-image-buffer-stores.ll:27]: No the old optimization is not wrong. GFX12 has an incompatible change in the behaviour of the instruction. The old optimization takes advantage of the pre-GFX12 behaviour, and the new optimization takes advantage of the GFX12 behaviour. This relies on the subtarget being set correctly so we know whether we're compiling for GFX12 or not - is that a problem?

Comment (jayfoad) [reply llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-simplify-image-buffer-stores.ll:27]: > We can't let optimizations run on IR without a set target-cpu break a program that ultimately executes on gfx12.

So is it OK for this optimization to do different things for GFX12 and pre-GFX12? Do we need a third option where it does nothing if the target CPU is not known?

Comment (arsenm) [reply llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-simplify-image-buffer-stores.ll:27]: So we need to definitively know if we're gfx12 or not. Ideally we would have a has-gfx12-behavior and has-pre-gfx12-behavior subtarget feature, and the combine would check both

**Review by jayfoad — APPROVED** [review]
LGTM. @arsenm does this address your concerns?

**Comment by mariusz-sikora-at-amd** [comment]
ping @arsenm

**Comment by mariusz-sikora-at-amd** [comment]
Merge with upstream to run tests. I will merge this changes if CI will pass.

File: llvm/lib/Target/AMDGPU/AMDGPU.td [hunk 2]
```diff
@@ -719,6 +719,18 @@ def FeatureFlatAtomicFaddF32Inst
...
+def FeatureDefaultComponentZero : SubtargetFeature<"default-component-zero",
+  "HasDefaultComponentZero",
+  "true",
+  "BUFFER/IMAGE store instructions set unspecified components to zero"
+>;
+
+def FeatureDefaultComponentBroadcast : SubtargetFeature<"default-component-broadcast",
+  "HasDefaultComponentBroadcast",
+  "true",
+  "BUFFER/IMAGE store instructions set unspecified components to x component"
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPU.td:731]: Probably should add gfx numbers to the description

**Pull request merged**